### ISSUE TITLE
TDNR doesn’t work for higher-rank functions

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#2381]
So if I have

```
Scope.run : (forall s . ‘{Scope s} a) -> a
Blah.run  : (forall s . ‘{Blah s} a) -> a
```
And then I write `run` somewhere, it won’t be able to resolve to `Scope.run` or `Blah.run`.

The workaround is to refer to `Scope.run` or `Blah.run` explicitly.

**Comment by dolio** [comment]
Is there an example for where it doesn't work?

My guess would be that this is akin to expecting higher-rank types to be inferred automatically, because a higher rank type needs to be inferred for the higher rank candidates to be subtypes. That is not a tractable expectation.

**Comment by pchiusano** [comment]
> My guess would be that this is akin to expecting higher-rank types to be inferred automatically, because a higher rank type needs to be inferred for the higher rank candidates to be subtypes. That is not a tractable expectation.

Yeah, I think it's that.

I could imagine TDNR working differently though and doing checking rather than inference. The way it works right now is something like - it infers the type of the symbol from context, then sees if that matches the types of symbols with a matching suffix. That approach obviously won't find higher-rank stuff since that will never be inferred.

Does not seem high priority though.

**Comment by dolio** [comment]
Yeah, I'm not sure how to fix it other than just trying to type check with all possibilities.
